crates/orderbook/src/database/trades.rs

```diff
@@ -25,27 +25,56 @@ pub struct TradeFilter {
 #[async_trait::async_trait]
 impl TradeRetrieving for Postgres {
     async fn trades(&self, filter: &TradeFilter) -> Result<Vec<Trade>> {
-        let _timer = super::Metrics::get()
+        let timer = super::Metrics::get()
             .database_queries
             .with_label_values(&["trades"])
             .start_timer();
 
         let mut ex = self.pool.acquire().await?;
-        database::trades::trades(
+        let trades = database::trades::trades(
             &mut ex,
             filter.owner.map(|owner| ByteArray(owner.0)).as_ref(),
             filter.order_uid.map(|uid| ByteArray(uid.0)).as_ref(),
         )
-        .map(|result| match result {
-            Ok(row) => trade_from(row),
-            Err(err) => Err(anyhow::Error::from(err)),
-        })
-        .try_collect()
+        .map_err(anyhow::Error::from)
+        .try_collect::<Vec<TradesQueryRow>>()
+        .await?;
+        timer.stop_and_record();
+
+        let order_uids = trades.iter().map(|t| t.order_uid).collect::<Vec<_>>();
+        let timer = super::Metrics::get()
+            .database_queries
+            .with_label_values(&["order_quotes"])
+            .start_timer();
+        let quotes = database::orders::read_quotes(&mut ex, order_uids.as_slice())
+            .await?
+            .into_iter()
+            .map(|quote| (quote.order_uid, quote))
+            .collect::<HashMap<_, _>>();
+        timer.stop_and_record();
+
+        try_join_all(
+            trades
+                .into_iter()
+                .map(|trade| {
+                    let quote = quotes.get(&trade.order_uid);
+                    async move {
+                        match trade.auction_id {
+                            Some(auction_id) => {
+                                self.fee_policies(auction_id, trade.order_uid, quote).await
```

Comment:
Can this be a single roundtrip (fetch fee policies for a list of `auction/order` pairs) and then only for the ones that require a quote we fetch the quote (in another roundtrip)?

Reply:
Replaced with a single roundtrip.

Reply:
Replaced with a single roundtrip.